pkg/controllers/infracluster/azure.go

```diff
@@ -0,0 +1,238 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package infracluster
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/url"
+	"strconv"
+
+	"github.com/go-logr/logr"
+	mapiv1beta1 "github.com/openshift/api/machine/v1beta1"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ptr "k8s.io/utils/ptr"
+	azurev1 "sigs.k8s.io/cluster-api-provider-azure/api/v1beta1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1beta1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/yaml"
+)
+
+var (
+	errUnableToGetAzureMAPIProviderSpec = errors.New("unable to get Azure MAPI ProviderSpec")
+	errAzureStackNotSupported           = errors.New("AzureStack Platform Type not yet supported")
+)
+
+const (
+	clusterSecretName               = "capz-manager-cluster-credential"    // #nosec G101
+	capzManagerBootstrapCredentials = "capz-manager-bootstrap-credentials" // #nosec G101
+)
+
+// ensureAzureCluster ensures the AzureCluster cluster object exists.
+//
+//nolint:funlen
+func (r *InfraClusterController) ensureAzureCluster(ctx context.Context, log logr.Logger) (client.Object, error) {
+	target := &azurev1.AzureCluster{ObjectMeta: metav1.ObjectMeta{
+		Name:      r.Infra.Status.InfrastructureName,
+		Namespace: defaultCAPINamespace,
+	}}
+
+	capzManagerBootstrapCredentialsKey := client.ObjectKey{Namespace: defaultCAPINamespace, Name: capzManagerBootstrapCredentials}
+	capzManagerBootstrapCredentials := &corev1.Secret{}
+
+	if err := r.Get(ctx, capzManagerBootstrapCredentialsKey, capzManagerBootstrapCredentials); err != nil {
+		return nil, fmt.Errorf("failed to get Azure Cluster Secret: %w", err)
+	}
+
+	err := r.Get(ctx, client.ObjectKeyFromObject(target), target)
+
+	switch {
+	case err != nil && !apierrors.IsNotFound(err):
+		return nil, fmt.Errorf("failed to get InfraCluster: %w", err)
+
+	case target.Spec.AzureEnvironment == "AzureStackCloud":
+		log.Info("AzureEnvironment not yet supported")
+		return nil, errAzureStackNotSupported
+
+	case target.GetAnnotations()[clusterv1.ManagedByAnnotation] == managedByAnnotationValueClusterCAPIOperatorInfraClusterController:
+		if err := r.Get(ctx, capzManagerBootstrapCredentialsKey, capzManagerBootstrapCredentials); err != nil {
+			return nil, fmt.Errorf("failed to get Azure Cluster Secret: %w", err)
+		} else if capzManagerBootstrapCredentialsKey.Namespace == "" || capzManagerBootstrapCredentialsKey.Name == "" {
+			return nil, fmt.Errorf("bootstrap credentials are not expected values: %w", err)
```

Comment:
It's not an error if they aren't the values we would populate, we just want to make sure that we only ensure the secret when the values are the values we would populate